### PR TITLE
New Item details UI : Part 1

### PR DESCRIPTION
**Pull request by patrixr**
## New Item view UI (Part 1)

Related ticket: https://jira.crossroads.org.hk/browse/GCW-2778

It is not a completed implementation of the [mockup](https://xd.adobe.com/view/224f7314-f1c0-40b6-7d7a-929e616be6bb-547e/screen/969910c2-f702-4f41-9c98-68ccd4741aa9/iPhone-6-7-8-85) , some more features will come in subsequent PRs.

### UI Additions

- A (default) `info` tab on the item page
- A `location` tab on the item page
- A `publishing` tab on the item page
- An overlay used to switch between items of the same set

### Other

- Creation of a new toggle-switch component 
- Creation of a new `matches` template helper
- Fixed a race condition where images would be loaded after the image carousel, causing it to freak out
- Fixed back links for the `search_location` to support the new tabs (todo: refactor, we can't have huge lists of `if/else` like this anymore)

### Screenshot

Screenshot 2019-09-19 at 11 54 11 AM

### CODE REVIEW

File: app/routes/items/detail.js [hunk 1]
```diff
@@ -10,23 +10,32 @@ export default AuthorizeRoute.extend({
...
-    if(model.get('isSet')) {
-      return Ember.RSVP.hash({
-        items: model.get('setItem.items').forEach(item => {
-          this.store.findRecord("item", item.get("id"), { reload: true });
-        })
-      });
+    if (!model.get("inventoryNumber")) {
+      this.get("transition").abort();
+      this.get("messageBox").alert(
+        "This item is not inventoried yet or has been marked as missing.",
```
Comment (swatijadhav) [inline app/routes/items/detail.js]: Should we move this message to translations?

File: app/routes/items/detail.js [hunk 2]
```diff
@@ -36,20 +45,87 @@ export default AuthorizeRoute.extend({
...
+  /**
+   * Loads an image if not available
+   *
+   * @param {String} id the image id
+   * @returns {Image} the image record
+   */
+  loadImage(id) {
+    const cachedImg = this.store.peekRecord("image", id);
+    if (cachedImg) {
+      return Ember.RSVP.resolve(cachedImg);
```
Comment (swatijadhav) [inline app/routes/items/detail.js]: Do we need `Ember.RSVP.resolve` here? or we can only return `cachedImg` value?

File: app/routes/items/detail.js (hunk 1 above)
Comment (patrixr) [reply app/routes/items/detail.js]: Good catch 👍

File: app/styles/templates/items/_detail.scss [hunk 3]
```diff
@@ -15,6 +22,140 @@
...
+      // Black tranparent header
+      .image-overlay {
+        background-color: rgba(black, 0.6);
+        position: absolute;
+        bottom: 0;
+        width: 100%;
+        min-height: 4rem;
+        padding-top: 1rem;
+        padding-left: 1rem;
+        padding-right: 1rem;
```
Comment (swatijadhav) [inline app/styles/templates/items/_detail.scss]: Can we combine this into `padding: 1rem 1rem 0` ?

File: app/routes/items/detail.js (hunk 2 above)
Comment (patrixr) [reply app/routes/items/detail.js]: Kinda depends on how the caller method uses the returned value.
But we can't make any assumptions, therefore the returned type should always be a promise.

But I will transform this into an `async` method, to clear the `resolve` (same result, better looking)

File: app/styles/templates/items/_detail.scss (hunk 3 above)
Comment (patrixr) [reply app/styles/templates/items/_detail.scss]: It's just a preference of mine, more descriptive. Plus, I can never remember the order in`padding: 1rem 1rem 0` 😆

**Comment by patrixr** [comment]
@swatijadhav Updated according to comments

File: app/controllers/items/detail.js [hunk 4]
```diff
@@ -78,6 +91,58 @@ export default Ember.Controller.extend(singletonItemDispatchToGcOrder, {
   ),
 
   actions: {
+    /**
+     * Called after a property is changed to push the updated
```
Comment (abulsayyad123) [inline app/controllers/items/detail.js]: Comments 👍

File: app/routes/application.js [hunk 5]
```diff
@@ -65,11 +67,14 @@ export default Ember.Route.extend(preloadDataMixin, {
...
   getErrorMessage(reason) {
-    if (reason.errors.length && reason.errors[0].detail && reason.errors[0].detail.status == 422) {
-      return reason.errors[0].detail.message;
-    } else {
-      return this.get("i18n").t("unexpected_error");
+    const status = _.get(reason, "errors[0].detail.status");
+    const defaultMessage = this.get("i18n").t("unexpected_error");
+
+    if (status === 422) {
+      return _.get(reason, "errors[0].detail.message", defaultMessage);
```
Comment (abulsayyad123) [inline app/routes/application.js]: 👌

File: app/routes/items/detail.js [hunk 6]
```diff
@@ -36,20 +46,87 @@ export default AuthorizeRoute.extend({
...
+   *
+   * @param {String} id the item id
+   * @param {Object} opts load options
+   * @param {Boolean} opts.loadImages if set to true, will also load the images
+   * @returns {Item} the item record
+   */
+  async loadItem(id, opts = {}) {
+    const { loadImages = false } = opts;
+
+    const it = await this.store.findRecord("item", id, { reload: true });
```
Comment (abulsayyad123) [inline app/routes/items/detail.js]: What does `it` refers to ?

Comment (patrixr) [reply app/routes/items/detail.js]: oh, it's meant as a shorter`item` , I can replace it if unclear.

I usually use it in loops  `items.forEach(it => { ... })` , but I guess it doesn't make too much sense here

Comment (patrixr) [reply app/routes/items/detail.js]: Updated 👌

File: app/helpers/matches.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+ * {{#if (matches myString '^(foo|bar)$')}}
+ *   It matches !
+ * {{/if}}
+ *
+ * @param {String} text the text to test
+ * @param {String} regex the regex to test the text against
+ * @returns {Boolean}
+ */
+export default Ember.Helper.helper(function([text, regex]) {
+  return new RegExp(regex).test(text);
```
Comment (namrataukirde) [inline app/helpers/matches.js]: Great 👍

**Comment by patrixr** [comment]
Merging now to start testing, there will still be some upcoming changes. So feel free to keep dropping messages on the PR.

**Pull request merged**